sdk/lib/opentelemetry/sdk/trace/export/batch_sampled_span_processor.rb

```diff
@@ -21,7 +21,90 @@ module Export
         # worker thread that exports the spans to wake up and start a new
         # export cycle.
         class BatchSampledSpanProcessor
-          # TODO
+          SCHEDULE_DELAY_MILLIS = 5000
+          MAX_QUEUE_SIZE = 2048
+          MAX_EXPORT_BATCH_SIZE = 512
+
+          def initialize(exporter:, schedule_delay: SCHEDULE_DELAY_MILLIS, max_queue_size: MAX_QUEUE_SIZE, max_export_batch_size: MAX_EXPORT_BATCH_SIZE)
+            @exporter = exporter
+            @mutex = Mutex.new
+            @condition = ConditionVariable.new
+            @keep_running = true
+            @delay = schedule_delay
+            @max_queue_size = max_queue_size
+            @batch_size = max_export_batch_size
+            @spans = []
+            @thread = Thread.new { work }
+          end
+
+          # does nothing for this processor
+          def on_start(span)
+            # noop
+          end
+
+          # adds a span to the batcher, threadsafe may block on lock
+          def on_end(span)
+            return unless span.recorded?
+
+            lock do
+              spans.shift if spans.size >= max_queue_size
+              spans << span
+              @condition.signal if spans.size > max_queue_size / 2
+            end
+          end
+
+          # shuts the consumer thread down and flushes the current accumulated buffer
+          # will block until the thread is finished
+          def shutdown
+            lock do
+              @keep_running = false
+              @condition.signal
+            end
+
+            @thread.join
+          end
+
+          private
+
+          attr_reader :spans, :max_queue_size, :batch_size
+
+          def work
+            loop do
+              keep_running = nil
+              batch = nil
+              lock do
+                if spans.size < max_queue_size
+                  @condition.wait(@mutex, @delay) while spans.empty? && @keep_running
+                  keep_running = @keep_running
+                end
+                batch = fetch_batch
+              end
+              # this is done outside the lock to unblock the producers
+              @exporter.export(batch)
+              break unless keep_running
+            end
+            flush
+          end
+
+          def flush
+            @exporter.export(fetch_batch) until spans.empty?
```

Comment:
This needs to lock while calling `fetch_batch`. It is also possible that the producers will keep producing indefinitely, outpacing the `flush`. It is preferable to take a snapshot of the `spans` and flush those, while ignoring any spans added after the snapshot.